audit/base/base.go

```diff
@@ -510,23 +554,42 @@ func (c *Client) write(ctx context.Context, msg msgs.Msg) {
 		c.setErr(fmt.Errorf("%w: %w", err, ErrConnection))
 		return
 	}
+	c.metrics.msgsSent.Add(ctx, 1)
 	c.successSend = true
 }
 
+// msgRequeueOrDrop requeues the message if there is room in the queue. If there is no room, the message is dropped.
+func (c *Client) msgRequeueOrDrop(ctx context.Context, msg msgs.Msg, err error) {
+	select {
+	case c.sendCh <- SendMsg{Ctx: ctx, Msg: msg}:
+		c.metrics.requeuedCounter.Add(1)
+		c.metrics.msgsRequeued.Add(ctx, 1)
+	default:
+		c.metrics.msgsDropped.Add(ctx, 1)
+		c.log.Error(fmt.Sprintf("audit message dropped due to queue being full: %v", err))
+	}
+}
+
 // setErr sets the error for the client if the error is not already set. If the error being set it nil,
 // it will be set to nil.
 func (c *Client) setErr(err error) {
 	c.setErrMu.Lock()
 	defer c.setErrMu.Unlock()
 
+	// This resets the error if the error passed is nil.
 	if err == nil {
 		c.err.Store(nil)
+		return
 	}
 
+	// If there already is an error, we don't need to overwrite it.
 	if c.err.Load() != nil {
 		return
 	}
 
+	// This sets the conn to nil. This prevents any further writes to the connection until it gets replaced.
+	c.conn.Store(nil)
```

Comment:
Doesn't this assume that the error is fatal? Are all errors fatal? 

Somewhat answering my own question: Other errors are defined and not clearly documented as fatal, but none are set on the client itself (i.e. `ValidationError` is never passed to `setErr`).

Open question, what should the behavior of the message writing be if something like `ErrQueueFull` were passed to `setErr` - I assume that's sorta nonsensical? Do we want to guard against it with an `IsFatal` before setting the conn to nil, or does it always make sense to set the conn to nil?

Reply:
It does and is only used in places where the error is fatal.  I've now documented that to make that clear.

So yea, ErrQueueFull is nonsensical for this purpose.  I don't think we should guard against it, because that leaves it open for someone doing just that, which would have other consequences.